File: src/AutoRest.SdkExplorer/Model/Azure/AzureResourceIdentifier.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

using System;
using System.Collections.Generic;
using System.Linq;

namespace AutoRest.SdkExplorer.Model.Azure
{
    public class AzureResourceIdentifier
    {
        // save the raw id only for readability and troubleshooting purpose
        public string RawId { get; set; }
        // segments and action are source of truth
        public List<AzureResourceIdentifierSegment> ResourceSegments { get; set; }
        public string Action { get; set; }

        public string GetId()
        {
            string resourcePart = string.Join("/", this.ResourceSegments.Select(s => s.ToString()));
            string actionPart = this.Action == null ? "" : this.Action;

            if (string.IsNullOrEmpty(resourcePart))
                return string.Empty;
            if (string.IsNullOrEmpty(actionPart))
                return $"/{resourcePart}";
            else
                return $"/{resourcePart}/{actionPart}";
        }
        public string? GetSubscriptionId() { return this.GetResourceTypeValue("subscriptions"); }
        public string? GetResourceGroupName() { return this.GetResourceTypeValue("resourcegroups"); }
        public AzureResourceType? GetResourceType() {
            int providerIndex = GetResourceTypeIndex("providers");
            if (providerIndex >= 0)
            {
                string ns = this.ResourceSegments[providerIndex].Value;
                string type = string.Join("/", this.ResourceSegments.Take(new Range(providerIndex + 1, this.ResourceSegments.Count)).Select(s => s.Type));
                return new AzureResourceType(ns, type);
            }
            return null;
        }
        public string? GetLastResourceName() { return this.ResourceSegments.Count > 0? this.ResourceSegments[^1].Value : null; }

        /// <summary>
        /// just for json/yaml ser/des
        /// </summary>
        private AzureResourceIdentifier()
        {
            this.RawId = string.Empty;
            this.ResourceSegments = new List<AzureResourceIdentifierSegment>();
            this.Action = string.Empty;
        }

        public AzureResourceIdentifier(string id)
        {
            this.RawId = id;
            this.ResourceSegments = new List<AzureResourceIdentifierSegment>();
            this.Action = string.Empty;
            this.Parse(id);
        }

        public string? GetResourceTypeValue(string resourceTypeName)
        {
            return this.ResourceSegments.FirstOrDefault(s => s.Type == resourceTypeName)?.Value;
        }

        public int GetResourceTypeIndex(string resourceTypeName)
        {
            int i = 0;
            foreach (var seg in this.ResourceSegments)
            {
                if (seg.Type == resourceTypeName)
                    return i;
                i++;
            }
            return -1;
        }

        private List<AzureResourceIdentifierSegment> CreateSegmentsList(string[] arr, out string action)
        {
            int segLen = arr.Length / 2;
            List<AzureResourceIdentifierSegment> segs = new List<AzureResourceIdentifierSegment>();
            for (int i = 0; i < segLen; i++)
            {
                segs.Add(new AzureResourceIdentifierSegment(arr[2 * i], arr[2 * i + 1]));
            }
            action = (arr.Length % 2 != 0) ? arr[^1] : string.Empty;
            return segs;
        }

        private void Parse(string id)
        {
            if (string.IsNullOrEmpty(id))
                throw new ArgumentNullException("id");

            string[] arr = id.Split('/', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries);

            if (arr.Length == 0)
                throw new ArgumentException("At least one segment is expected in id. Id=" + id);

            var segList = CreateSegmentsList(arr, out string action);
            this.ResourceSegments = segList;
            this.Action = action;
        }
    }
}

File: src/AutoRest.SdkExplorer.TypeScript/src/Model/Azure/AzureResourceIdentifier.ts
import { isStringEqualCaseInsensitive, isStringNullOrEmpty } from "../../Utils/utils";
import { AzureResource } from "./AzureResource";
import { AzureResourceIdentifierSegment } from "./AzureResourceIdentifierSegment";
import { AzureResourceType } from "./AzureResourceType";


export class AzureResourceIdentifier implements AutoRest.SdkExplorer.Interface.AzureResourceIdentifier {
    /**
     * only for readability and troubleshooting purpose, may be empty. Use getId() to get the real id
     */
    public rawId: string;
    public resourceSegments: AzureResourceIdentifierSegment[] = [];
    public action: string = "";

    public getId(): string {
        return this.generateId(this.resourceSegments.length, true);
    }

    public get isEmpty() { return isStringNullOrEmpty(this.getId()); }
    public get hasParameter() { return this.resourceSegments.findIndex(s => s.isParameterized) >= 0; }

    public getSubscriptionId(): string | undefined {
        return this.GetResourceTypeValue("subscriptions");
    }

    public getResourceGroupName(): string | undefined {
        return this.GetResourceTypeValue("resourcegroups");
    }

    public getResourceType(): AzureResourceType | undefined {
        const providerIndex = this.GetResourceTypeIndex("providers");
        if (providerIndex >= 0) {
            const ns = this.resourceSegments[providerIndex].value;
            const value = this.resourceSegments.slice(providerIndex + 1).map(v => v.type).join("/");
            return new AzureResourceType({ namespace: ns, type: value });
        }
        return undefined;
    }

    public getLastResourceName(): string | undefined {
        return this.resourceSegments.length > 0 ? this.resourceSegments[this.resourceSegments.length - 1].value : undefined;
    }

    private generateId(takeSegmentCount: number, includeAction: boolean): string {
        const resourcePart = this.resourceSegments.slice(0, takeSegmentCount).map(s => `${s.type}/${s.value}`).join("/");
        const actionPart = (!includeAction || this.action == null || this.action.length === 0) ? "" : `${this.action}`;

        if (resourcePart.length === 0)
            return "";
        if (actionPart.length === 0)
            return `/${resourcePart}`;
        else
            return `/${resourcePart}/${actionPart}`;
    }

    public getParentIdentifierAt(lastSegmentIndex: number): AzureResourceIdentifier | undefined {
        const id = this.generateId(lastSegmentIndex + 1, false);
        return AzureResourceIdentifier.Parse(id);
    }

    constructor(data: AutoRest.SdkExplorer.Interface.AzureResourceIdentifier) {
        this.action = data.action;
        this.rawId = data.rawId;
        this.resourceSegments = data.resourceSegments.map(s => new AzureResourceIdentifierSegment(s));
    }

    private GetResourceTypeValue(resourceTypeName: string): string | undefined {
        return this.resourceSegments.find(s => isStringEqualCaseInsensitive(s.type, resourceTypeName))?.value;
    }

    private GetResourceTypeIndex(resourceName: string): number {
        return this.resourceSegments.findIndex(s => isStringEqualCaseInsensitive(s.type, resourceName));
    }

    public static IsPatternMatch(idA: string | undefined, idB: string | undefined): boolean {
        if (isStringNullOrEmpty(idA) && isStringNullOrEmpty(idB))
            return true;
        if (isStringNullOrEmpty(idA) || isStringNullOrEmpty(idB))
            return false;
        const a = AzureResourceIdentifier.Parse(idA!);
        const b = AzureResourceIdentifier.Parse(idB!);

        if (a.resourceSegments.length !== b.resourceSegments.length)
            return false;
        if (!AzureResourceType.isEquals(a.getResourceType(), b.getResourceType()))
            return false;
        if (!isStringEqualCaseInsensitive(a.action, b.action))
            return false;
        return true;
    }

    // from id like  /subscriptions/{subscription-id}/resourceGroups/myResourceGroup/providers/Microsoft.Network/networkInterfaces/{existing-nic-name}
    public static Parse(idStrng: string): AzureResourceIdentifier {

        const arr = idStrng.split(/[/\\]/).map(v => v.trim()).filter(v => v.length > 0);

        let i = 1;
        var segs = [];
        var action = "";
        while (i < arr.length) {
            segs.push(new AzureResourceIdentifierSegment({ type: arr[i - 1], value: arr[i] }));
            i += 2;
        }
        if (i == arr.length)
            action = arr[i - 1];
        const r = new AzureResourceIdentifier({ rawId: idStrng, resourceSegments: segs, action: action });
        return r;
    }

    /**
     * 
     * @param idString - an resource id with parameterized resource name/value wrapped with {}, i.e. request path like: 
     *              "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/privateDnsZones/{privateZoneName}"
     * @param resource - replace parameters with the value from the resource if possible
     * @returns
     */
    public static ParseWithResource(idString: string, resource: AzureResource): AzureResourceIdentifier {

        const pAri = AzureResourceIdentifier.Parse(idString)
        const resAri = AzureResourceIdentifier.Parse(resource.id);

        pAri.resourceSegments.forEach(seg => {
            if (seg.isParameterized) {
                const found = resAri.GetResourceTypeValue(seg.type);
                if (found)
                    seg.value = found;
            }
        });

        return pAri;
    }
}

